### PR TITLE
Improve seed search

### PR DESCRIPTION
**Pull request by badnames**
This PR addresses the first bullet point from #209 

## Basics

- [x] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [x] The buildserver is happy.
- [x] The PR is rebased with current master.

## Checklist

- [ ] I added unit tests for my code
- [x] I fully described what my PR does in the documentation
      (not in the PR description)
- [x] I fixed all affected documentation
- [ ] I fixed all affected decisions
- [x] I added code comments, logging, and assertions as appropriate
- [ ] I mentioned [every code or binary](/.reuse/dep5) not directly written or done by me in [reuse syntax](https://reuse.software/)

## Review

- [ ] Documentation is conforming to [our Documentation Guidelines](/doc/documentation.md)
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to our Coding Guidelines
- [ ] Code is consistent to [our Design Decisions](/doc/decisions)

### CODE REVIEW

**Comment by badnames** [comment]
I still have to get our custom styles to work with PaginatedSelectMenu tomorrow, everything else should be ready for review.

File: backend/src/service/plants.rs [hunk 1]
```diff
@@ -30,7 +32,13 @@ pub async fn search(
     query: &PlantsSearchParameters,
 ) -> Result<Vec<PlantsSummaryDto>, ServiceError> {
     let mut conn = pool.get().await?;
-    let result = Plants::search(query, &mut conn).await?;
+
+    // pages start at 1
+    let calculated_offset = query.limit * (query.page - 1);
+    // dissalow negative offsets
```
Comment (markus2330) [inline backend/src/service/plants.rs:38]:
```suggestion
    // disallows negative offsets
```

File: backend/src/service/plants.rs [hunk 2]
```diff
@@ -30,7 +32,13 @@ pub async fn search(
     query: &PlantsSearchParameters,
 ) -> Result<Vec<PlantsSummaryDto>, ServiceError> {
     let mut conn = pool.get().await?;
-    let result = Plants::search(query, &mut conn).await?;
+
+    // pages start at 1
+    let calculated_offset = query.limit * (query.page - 1);
```
Comment (markus2330) [inline backend/src/service/plants.rs:37]:
```suggestion
    let pages_start_at_1 = 1;
    let calculated_offset = query.limit * (query.page - pages_start_at_1);
```

File: frontend/src/components/Form/PaginatedSelectMenu.tsx [hunk 3]
```diff
@@ -0,0 +1,110 @@
+import filterObject from '@/utils/filterObject';
+import { Control, Controller, FieldValues, Path } from 'react-hook-form';
+import StylesConfig from 'react-select';
+import AnycPaginate, { AsyncPaginate } from 'react-select-async-paginate';
+import { ClassNamesConfig } from 'react-select/dist/declarations/src/styles';
+import { SelectOption } from './SelectMenu';
+
+/**
+ * Contains the information needed by react-select-async-paginate for loading a single page.
+* See https://github.com/vtaits/react-select-async-paginate/tree/master/packages/react-select-async-paginate#loadoptions for more information.
```
Comment (markus2330) [inline frontend/src/components/Form/PaginatedSelectMenu.tsx:10]:
```suggestion
 * See https://github.com/vtaits/react-select-async-paginate/tree/master/packages/react-select-async-paginate#loadoptions for more information.
```

**Review by markus2330 — CHANGES_REQUESTED** [review]
Great job, pagination works for me. Please rebase/merge to master for the small fix needed for adding seeds to actually work.

File: backend/src/model/dto.rs [hunk 4]
```diff
@@ -63,10 +63,15 @@ pub struct PlantsSummaryDto {
     pub common_name: Option<Vec<Option<String>>>,
 }
 
-#[allow(clippy::missing_docs_in_private_items)] // TODO: See #97.
+/// Query parameters for the searching plants.
 #[typeshare]
 #[derive(Debug, Deserialize)]
 pub struct PlantsSearchParameters {
+    /// The system will check if this string occurs in the plants common name or binomial name.
```
[the remark this answers is not in the document]
Comment (badnames) [reply backend/src/model/dto.rs:70]: I have thought a bit about this over the past few days and came to the conclusion that we probably shouldn't search by German and English common names at the same time.

As a user, I would personally find it very confusing if I started typing "fi" and suddenly spruces (or "Fichten" in German) showed up in the search results, when I actually wanted to find firs.
In my opinion it would probably be better to add some mechanism to select which type of common name should be searched for. Preferably depending on the selected language.

If you want us to implement the search this way none the less we should at least show the German common name beside the English common name.

[the remark this answers is not in the document]
Comment (badnames) [reply backend/src/model/dto.rs:70]: But, whathever we choose to do should certainly be part of its own PR to avoid the same mistake we made in the last PR for the seed use case....

[the remark this answers is not in the document]
Comment (markus2330) [reply backend/src/model/dto.rs:70]: I definitely agree about not putting everything in one PR, I think our feedback cycle is fast enough that this is not a problem. Probably I can already merge already now, or is something missing here?

The thing about German/English is that sometimes I remember the German name, sometimes the English. Some plants are quite exotic, the German name might be quite unknown/unusual. E.g. for [Pawpaw](https://de.wikipedia.org/wiki/Dreilappige_Papau) in principle has has the German names Indianerbanane or Papau but many people in Austria (including Yvonne and me) call them Pawpaw. Of course we could add "Pawpaw" also to German common names but this would be extra effort in maintaining the plants database.

Furthermore, I thought that a solution where we search everything at once makes the functionality (or at least the API) easier. Is this assumption wrong?

> As a user, I would personally find it very confusing if I started typing "fi" and suddenly spruces (or "Fichten" in German) showed up in the search results, when I actually wanted to find firs.

What if we simply rank English names in English settings higher? (When we implement the ranking.) Full matches also should be very high in ranking, so when typing "fir", "fir" definitely should be the first match, see [update of use case](https://github.com/ElektraInitiative/PermaplanT/pull/251/files#diff-ffcf3684203226512e824186cca963094610c83f56ecdffd18393acf5c281600)

**Review by Bushuo — COMMENTED** [review]
Really nice :)

File: frontend/src/features/seeds/api/searchPlants.ts [hunk 5]
```diff
@@ -2,16 +2,30 @@ import { PlantsSummaryDto } from '@/bindings/definitions';
...
 
     const response = await axios.get<PlantsSummaryDto[]>(
       `${baseApiUrl}/api/plants?${searchParams}`,
     );
-    return response.data;
+    
+    return {
+      dtos: response.data,
+      // If a page could not be filled completely, there will propably be no more data.
+      hasMore: response.data.length == 10,
```
[the remark this answers is not in the document]
Comment (badnames) [reply frontend/src/features/seeds/api/searchPlants.ts:27]: Yes, in the edge case you mentioned the backend would just return an empty page for the last request.  The frontend would then pass said empty page to the select menu component with the hasMore flag set to false.
So in short this would just result in a single unnecessary request.    

Loading an additional page on each request would require some state management to avoid doubling the network traffic. 

So in short I chose to do it this way to keep the code nice and simple, while avoiding additional  requests.

[the remark this answers is not in the document]
Comment (Bushuo) [reply frontend/src/features/seeds/api/searchPlants.ts:27]: > Loading an additional page on each request would require some state management to avoid doubling the network traffic.

I did not mean to load an additional page. Just one more item per request in the backend database query.
```rs
...
.limit(limit.into() + 1)
...

// something like this
let has_more = query_result.len() == limit + 1
```
but it does not matter anyway. The UX feels good.

File: backend/src/model/dto.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (badnames) [reply backend/src/model/dto.rs:70]: > The thing about German/English is that sometimes I remember the German name, sometimes the English. Some plants are quite exotic, the German name might be quite unknown/unusual. E.g. for [Pawpaw](https://de.wikipedia.org/wiki/Dreilappige_Papau) in principle has has the German names Indianerbanane or Papau but many people in Austria (including Yvonne and me) call them Pawpaw. Of course we could add "Pawpaw" also to German common names but this would be extra effort in maintaining the plants database.

Fair point. 
However: shouldn't "Pawpaw" then also be considered a German common name, since its frequently used to refer to a that specific plant even though its a word of foreign origin?   
"Physalis" is also commonly used to refer to a ["Blasenkirsche"](https://de.wikipedia.org/wiki/Blasenkirschen) for example.
Before finding the linked Wikipedia article I wasn't even aware that it had a "proper" German name to be honest.

And even if you can't remember the name of a plant in your preferred language, is it too much to ask of the user to either change the interface language or do 30 second lookup in an online dictionary or wikipedia?

Again, I don't want to sound like I'm trying to avoid work here but this idea is again something that

a. Might be more complicated to implement than expected.
b. Potentially has a very negative impact on user experience. 
c. Solves a rather specific problem that can easily be overcome using external tools.

This is why I'm of the opinion that the risks are not worth the rather small reward.

> Furthermore, I thought that a solution where we search everything at once makes the functionality (or at least the API) easier. Is this assumption wrong?

Yes, I think so.
If we go with the ranking method you described we would still have to communicate the users preferred language to the backend so that we know how the results should be ranked. 
This is exactly the same information needed to determine which column should be used by the search if we limit go with the "to one language at a time" approach.

File: frontend/src/features/seeds/components/CreateSeedForm.tsx [hunk 6]
```diff
@@ -43,16 +40,40 @@ const CreateSeedForm = ({
     onSubmit(data);
   };
 
+  const loadPlants = async (search: string, options: unknown, additional: PageAdditionalInfo | undefined): Promise<Page> => {
+    const page = additional != undefined ? additional.page : 1;
+    
+    console.log(page);
```
Comment (Bushuo) [inline frontend/src/features/seeds/components/CreateSeedForm.tsx:46]:
```suggestion
```

**Review by Bushuo — CHANGES_REQUESTED** [review]
I tested it a bit more and think that there is still a little bug.
If you open the seed entry page and tap the select menu. The first and only request is
`http://localhost:8080/api/plants?search_term=&limit=10&page=2`

If you write `tomato` all the request are of the form
`http://localhost:8080/api/plants?search_term=......&limit=10&page=2`

it always starts at page 2.

File: frontend/src/features/seeds/api/searchPlants.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (badnames) [reply frontend/src/features/seeds/api/searchPlants.ts:27]: Ok sorry, I encountered some brain-lag there :sweat_smile:.
Forget what I said, this is an absolutely great suggestion.

**Comment by badnames** [comment]
@Bushuo Huh weird. I will look into it, thanks.

**Review by Bushuo — APPROVED** [review]

File: backend/src/model/dto.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (markus2330) [reply backend/src/model/dto.rs:80]: > However: shouldn't "Pawpaw" then also be considered a German common name anyway, since its frequently used to refer to a that specific plant even though its a word of foreign origin?

As said, it is simply a huge effort to do that. It is better to improve the search than to duplicate many entries in the database.

> Still, if you can't remember the name of a plant in your preferred language, is it too much to ask of the user to either change the interface language or do 30 second lookup in an online dictionary or wikipedia?

This doesn't help you, as you would need to exactly know what is in **our database under which language**. When we are not able to present a match, it would lead to frustration.

> a. Might be more complicated to implement than expected, delaying the first milestone even further.

If there is something specifically which is complicated, please tell.

> b. Potentially has a very negative impact on user experience.

Presenting no match definitely is the worst user experience. This is also why search machines nowadays more or less ignore what you type and present you matches anyway.

> c. Solves a rather specific problem that can easily be overcome using external tools.

As said: unfortunately this is **not** the case. Yes, you could look up the German name ["Blasenkirsche](https://de.wikipedia.org/wiki/Blasenkirschen), but what if we **don't** have this name but have Andenkirsche instead? Or Kapstachelbeere? There is no external tool telling you what a person needs to type in our search field so that she will get results.

I updated #251, I hope it is fine now.

File: backend/src/model/dto.rs [hunk 7]
```diff
@@ -63,10 +63,25 @@ pub struct PlantsSummaryDto {
...
-#[allow(clippy::missing_docs_in_private_items)] // TODO: See #97.
+/// An array of PlantsSummaryDto with some basic pagination information used by the plants search endpoint.
+#[typeshare]
+#[derive(Debug, Serialize, PartialEq, Eq, Deserialize, ToSchema)]
+pub struct PlantsSearchDto {
+    /// A single page of searched plants.
+    pub plants: Vec<PlantsSummaryDto>,
+    /// Flag indicating whether there is more data on the next page.
+    pub has_more: bool,
+}
```
Comment (Benjaminimal) [inline backend/src/model/dto.rs:74]: We will have to solve this problem of not loading everything at once for other list endpoints as well.
I think we should generalize this into reusable offset-based pagination.
This would not only make things more uniform not only within this endpoint (response body is a list without params but an object with) but also for the whole API.

@markus2330 
I can create an issue and work on it in the coming week as my other work is probably blocked by the new plant columns anyway.

File: backend/src/controller/plants.rs [hunk 8]
```diff
@@ -16,19 +16,23 @@ use actix_web::{
...
+    let response = if let Some(parameters) = query {
+        let response = service::plants::search(&pool, &parameters).await?;
+        HttpResponse::Ok().json(response)
+    } else {
+        let response = service::plants::find_all(&pool).await?;
+        HttpResponse::Ok().json(response)
     };
-    Ok(HttpResponse::Ok().json(response))
+
+    Ok(response)
```
Comment (Benjaminimal) [inline backend/src/controller/plants.rs:35]: This looks a little awkward which is due to differing types returned from `find_all` and `search`.
The generated openapi documentation is also wrong now.
I would go with a more generic pagination approach.
See other comment.

File: backend/src/model/dto.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (badnames) [reply backend/src/model/dto.rs:80]: Ok, before I say anything else: I will go ahead and implement this in another PR as outlined in the new usecase over the course of the next week.

Just to explain my reasoning a bit:
> As said, it is simply a huge effort to do that.

Absolutely! I'm not arguing against that.  I meant to say that if *any kind of term* is commonly used to refer to a plant in German, it should *by definition* already be in the database as a common name in that language. If that isn't the case in the majority of cases we should probably reevaluate where we are getting our data from.

> If there is something specifically which is complicated, please tell.

The thing is, I have never worked on a search feature before (except maybe a few trivial ones), let alone in rust with actix web.
So its nothing specific really.
However, getting the ranking right could be a bit tricky (E.g. preventing elements from appearing twice in the search and so on).

> Presenting no match definitely is the worst user experience.

Maybe. Generally I would agree, but I think we all know about the "Kaufhaus Österreich" debacle where random stuff appeared in the product search. That played a big part in making the site borderline unusable.

Although I'd say this could be easily prevented in our case by making it obvious why a certain item was included in the search results. E.g. by highlighting the search term.

> As said: unfortunately this is not the case.

Again, debatable. For example: I don't think many non-German English speakers would benefit from search results coming from the German common name.  In the end it all comes down to who our user base is and what they expect.

> Yes, you could look up the German name ["Blasenkirsche](https://de.wikipedia.org/wiki/Blasenkirschen), but what if we don't have this name but have Andenkirsche instead? Or Kapstachelbeere? There is no external tool telling you what a person needs to type in our search field so that she will get results.

If all else fails you should hopefully be able to find the Latin name of a plant using the common name you know best.

File: backend/src/model/dto.rs (hunk 7 above)
Comment (badnames) [reply backend/src/model/dto.rs:74]: > I can create an issue and work on it in the coming week as my other work is probably blocked by the new plant columns anyway.

Yes please. I would prefer if we did this in another PR to keep this one at a manageable size.

Comment (markus2330) [reply backend/src/model/dto.rs:74]: Yes, great offer! :heart:

**Review by Benjaminimal — APPROVED** [review]

File: backend/src/model/dto.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Benjaminimal) [reply backend/src/model/dto.rs:80]: Turned it into its own issue.
https://github.com/ElektraInitiative/PermaplanT/issues/258
